File: android/app/src/main/java/com/rajarsheechatterjee/EpubUtil/EpubUtil.java
package com.rajarsheechatterjee.EpubUtil;

import android.util.Xml;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableArray;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.bridge.WritableNativeArray;
import com.facebook.react.bridge.WritableNativeMap;

import org.xmlpull.v1.XmlPullParser;
import org.xmlpull.v1.XmlPullParserException;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.util.HashMap;
import java.util.Objects;

public class EpubUtil extends ReactContextBaseJavaModule {
    EpubUtil(ReactApplicationContext context){super(context);}

    @NonNull
    @Override
    public String getName() {
        return "EpubUtil";
    }

    private XmlPullParser initParse(File file) throws XmlPullParserException, IOException {
        XmlPullParser parser = Xml.newPullParser();
        parser.setFeature(XmlPullParser.FEATURE_PROCESS_NAMESPACES, true);
        parser.setInput(new FileInputStream(file), null);
        parser.nextTag();
        return parser;
    }

    private String readText(XmlPullParser parser) throws IOException, XmlPullParserException {
        String result = "";
        if (parser.next() == XmlPullParser.TEXT) {
            result = parser.getText();
            parser.nextTag();
        }
        return result;
    }

    private String cleanUrl(String url){
        if(url != null){
            return url.replaceFirst("#[^.]+?$", "");
        }
        return null;
    }

    @ReactMethod
    public void parseNovelAndChapters(String epubDirPath, Promise promise) {
        try {
            File containerFile = new File(epubDirPath, "META-INF/container.xml");
            File contentMetaFile = new File(epubDirPath, getContentMetaFilePath(containerFile));
            String contentDir = contentMetaFile.getParent();

            ReadableMap novel = getNovelMetadata(contentMetaFile, contentDir);
            promise.resolve(novel);
        } catch (XmlPullParserException | IOException e) {
            promise.reject(e.getCause());
        }
    }

    private String getContentMetaFilePath(File file) throws XmlPullParserException, IOException {
        XmlPullParser parser = initParse(file);
        while (parser.next() != XmlPullParser.END_TAG){
            @Nullable String tag = parser.getName();
            if(tag != null && tag.equals("rootfile")) {
                return parser.getAttributeValue(null, "full-path");
            }
        }
        return "OEBPS/content.opf"; // default
    }
    private ReadableMap getNovelMetadata(File file, String contentDir) throws XmlPullParserException, IOException {
        WritableMap novel = new WritableNativeMap();
        WritableArray chapters = new WritableNativeArray();
        XmlPullParser parser = initParse(file);
        HashMap<String, String> refMap = new HashMap<>();
        HashMap<String, String> tocMap = new HashMap<>();
        File tocFile = new File(contentDir, "toc.ncx");
        if(tocFile.exists()){
            XmlPullParser tocParser = initParse(tocFile);
            String label = null;
            while (tocParser.next() != XmlPullParser.END_DOCUMENT){
                String tag = tocParser.getName();
                if(tag != null){
                    if(tag.equals("text")){
                        label = readText(tocParser);
                    }else if(tag.equals("content")){
                        String href = cleanUrl(tocParser.getAttributeValue(null, "src"));
                        if(href != null){
                            tocMap.put(href, label);
                        }
                    }
                }
            }
        }
        String cover = null;
        while (parser.next() != XmlPullParser.END_DOCUMENT){
            @Nullable String tag = parser.getName();
            if(tag != null){
                switch (tag) {
                    case "item": {
                        String id = parser.getAttributeValue(null, "id");
                        String href = parser.getAttributeValue(null, "href");
                        if (id != null) {
                            refMap.put(id, href);
                        }
                        break;
                    }
                    case "itemref": {
                        String idRef = parser.getAttributeValue(null, "idref");
                        String href = refMap.get(idRef);
                        if (href != null) {
                            WritableMap chapter = new WritableNativeMap();
                            chapter.putString("path", contentDir + "/" + href);
                            String name = tocMap.get(href);
                            chapter.putString("name", name == null ? href : name);
                            chapters.pushMap(chapter);
                        }
                        break;
                    }
                    case "title":
                        novel.putString("name", readText(parser));
                        break;
                    case "creator":
                        novel.putString("author", readText(parser));
                        break;
                    case "contributor":
                        novel.putString("artist", readText(parser));
                        break;
                    case "description":
                        novel.putString("summary", readText(parser));
                        break;
                    case "meta":
                        String metaName = parser.getAttributeValue(null, "name");
                        if(metaName != null && metaName.equals("cover")){
                            cover = parser.getAttributeValue(null, "content");
                        }
                        break;
                }
                parser.next();
            }
        }
        if(cover != null){
            String coverPath = contentDir + "/" + refMap.get(cover);
            novel.putString("cover", coverPath);
        }
        novel.putArray("chapters", chapters);
        return novel;
    }
}

File: android/app/src/main/java/com/rajarsheechatterjee/EpubUtil/EpubUtil.kt
package com.rajarsheechatterjee.EpubUtil

import android.util.Xml
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableArray
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.WritableArray
import com.facebook.react.bridge.WritableMap
import com.facebook.react.bridge.WritableNativeArray
import com.facebook.react.bridge.WritableNativeMap
import org.xmlpull.v1.XmlPullParser
import java.io.File
import java.io.FileInputStream

class EpubUtil(context: ReactApplicationContext) : ReactContextBaseJavaModule(context) {
    override fun getName(): String {
        return "EpubUtil"
    }

    private fun initParse(file: File): XmlPullParser {
        val parser = Xml.newPullParser()
        parser.setFeature(XmlPullParser.FEATURE_PROCESS_NAMESPACES, true)
        parser.setInput(FileInputStream(file), null)
        parser.nextTag()
        return parser
    }

    private fun readText(parser: XmlPullParser): String {
        var result = ""
        if (parser.next() == XmlPullParser.TEXT) {
            result = parser.text
            parser.nextTag()
        }
        return result
    }

    private fun cleanUrl(url: String): String {
        return url.replaceFirst("#[^.]+?$".toRegex(), "")
    }

    @ReactMethod
    fun parseNovelAndChapters(epubDirPath: String, promise: Promise) {
        try {
            val containerFile = File(epubDirPath, "META-INF/container.xml")
            val contentMetaFile = File(epubDirPath, getContentMetaFilePath(containerFile))
            val contentDir = contentMetaFile.parent
            val novel = contentDir?.let { getNovelMetadata(contentMetaFile, it) }
            promise.resolve(novel)
        } catch (e: Exception) {
            promise.reject(e)
        }
    }

    private fun getContentMetaFilePath(file: File): String {
        val parser = initParse(file)
        while (parser.next() != XmlPullParser.END_TAG) {
            val tag = parser.name
            if (tag != null && tag == "rootfile") {
                return parser.getAttributeValue(null, "full-path")
            }
        }
        return "OEBPS/content.opf" // default
    }

    private fun mergeChapters(
        entryList: List<ChapterEntry>,
        fileList: List<String>,
        contentDir: String
    ): ReadableArray {
        val foundIndexes = entryList.mapIndexed { index, entry ->
            val foundIndex = fileList.indexOf(entry.href)
            if (foundIndex < index) -1
            else foundIndex
        }
        val linkedIndex = foundIndexes.mapIndexed { arrayIndex, foundIndex ->
            if (foundIndex == -1) Pair(-1, -1)
            else {
                var nextIndex = -1
                for (j in arrayIndex + 1 until foundIndexes.size) {
                    if (foundIndexes[j] != -1) {
                        nextIndex = foundIndexes[j]
                        break
                    }
                }
                if (nextIndex == -1) nextIndex = fileList.size
                Pair(foundIndex, nextIndex)
            }
        }

        val chapters: WritableArray = WritableNativeArray()
        linkedIndex.forEachIndexed { arrayIndex, (firstIndex, lastIndex) ->
            val chapter = WritableNativeMap()
            val entryPath = "${contentDir}/${entryList[arrayIndex].href}"
            chapter.putString("name", entryList[arrayIndex].name)
            if (firstIndex == -1 || firstIndex == lastIndex) {
                chapter.putString("path", entryPath)
            } else {
                val mergedChapterPath = "${entryPath}-copy-$arrayIndex.html"
                chapter.putString("path", mergedChapterPath)
                val entryFile = File(entryPath)
                val mergedFile = File(mergedChapterPath)
                mergedFile.createNewFile()
                mergedFile.appendBytes(entryFile.readBytes())
                for (i in firstIndex + 1 until lastIndex) {
                    val file = File("${contentDir}/${fileList[i]}")
                    mergedFile.appendBytes(file.readBytes())
                }
            }
            chapters.pushMap(chapter)
        }
        return chapters
    }

    private fun getNovelMetadata(file: File, contentDir: String): ReadableMap {
        val novel: WritableMap = WritableNativeMap()
        val parser = initParse(file)
        val refMap = HashMap<String, String>()
        val entryList = mutableListOf<ChapterEntry>()
        val fileList = mutableListOf<String>()
        val tocFile = File(contentDir, "toc.ncx")
        if (tocFile.exists()) {
            val tocParser = initParse(tocFile)
            var label = ""
            while(tocParser.next() != XmlPullParser.END_DOCUMENT){
                if(tocParser.name == "navMap") break;
            }
            while (tocParser.next() != XmlPullParser.END_DOCUMENT) {
                val tag = tocParser.name
                if(tag == "navMap") break;
                if (tag != null) {
                    if (tag == "text") {
                        label = readText(tocParser)
                    } else if (tag == "content") {
                        val href = cleanUrl(tocParser.getAttributeValue(null, "src"))
                        if (label.isNotBlank() && File("${contentDir}/$href").exists()) {
                            entryList.add(ChapterEntry(name = label, href = href))
                        }
                        label = ""
                    }
                }
            }
        } else {
            throw Error("Table of content doesn't exist!")
        }
        var cover: String? = null
        while (parser.next() != XmlPullParser.END_DOCUMENT) {
            val tag = parser.name
            if (tag != null) {
                when (tag) {
                    "item" -> {
                        val id = parser.getAttributeValue(null, "id")
                        val href = parser.getAttributeValue(null, "href")
                        if (id != null) {
                            refMap[id] = href
                        }
                    }

                    "itemref" -> {
                        val idRef = parser.getAttributeValue(null, "idref")
                        val href = refMap[idRef]
                        if (href != null && File("$contentDir/$href").exists()) {
                            fileList.add(href)
                        }
                    }

                    "title" -> novel.putString("name", readText(parser))
                    "creator" -> novel.putString("author", readText(parser))
                    "contributor" -> novel.putString("artist", readText(parser))
                    "description" -> novel.putString("summary", readText(parser))
                    "meta" -> {
                        val metaName = parser.getAttributeValue(null, "name")
                        if (metaName != null && metaName == "cover") {
                            cover = parser.getAttributeValue(null, "content")
                        }
                    }
                }
                parser.next()
            }
        }
        val chapters = mergeChapters(entryList, fileList, contentDir)
        novel.putArray("chapters", chapters)
        if (cover != null) {
            val coverPath = contentDir + "/" + refMap[cover]
            novel.putString("cover", coverPath)
        } else {
            // try scanning Images dir if exists
            val imageDir = File("$contentDir/Images")
            if (imageDir.exists() && imageDir.isDirectory) {
                imageDir.listFiles()?.forEach { img ->
                    if (img.isFile && img.name.lowercase().contains("cover|illus".toRegex())) {
                        cover = img.path
                    }
                }
            }
            if (cover != null) {
                novel.putString("cover", cover)
            }
        }
        return novel
    }
}
